File: src/server/GraphController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.ModelBinding;
using Microsoft.FSharp.Collections;
using Microsoft.Quantum.Simulation.Simulators;
using static aleph.kets;
namespace aleph.server;

[ApiController]
[Route("[controller]")]
public class GraphController : ControllerBase
{
    private readonly IGraphsService _graphs;
    private readonly ILogger<GraphController> _logger;

    public GraphController(ILogger<GraphController> logger, IGraphsService graphs)
    {
        _logger = logger;
        _graphs = graphs;
    }

    [HttpGet()]
    public IEnumerable<QuantumGraph> GetMany()
    {
        return Enumerable.Empty<QuantumGraph>();
    }

    [HttpGet("~create")]
    [HttpPost("")]
    public string Create()
    {
        var id = Guid.NewGuid().ToString("D");
        var g = new QuantumGraph();
        _graphs.Add(id, g);
        return id;
    }

    [Route("{graphId}")]
    public ActionResult<QuantumGraph> GetGraph(string graphId)
    {
        if (_graphs.TryFind(graphId, out var g))
        {
            return g;
        }

        return NotFound();
    }

    [Route("{graphId}/{ketid}")]
    public ActionResult<GraphNode> GetNode(string graphId, int ketId)
    {
        if (_graphs.TryFind(graphId, out var graph))
        {
            return new GraphNode(ketId, graph);
        }

        return NotFound();
    }

    [Route("{graphId}/literal")]
    public ActionResult<int> Literal(string graphId, [BindRequired] int width) =>
        AddExpression(graphId, Expression.NewLiteral(width));

    [Route("{graphId}/bool")]
    public ActionResult<int> ConstantBool(string graphId, [BindRequired] bool value) =>
        AddExpression(graphId, Expression.NewConstant(value ? 1 : 0));

    [Route("{graphId}/int")]
    public ActionResult<int> ConstantInt(string graphId, [BindRequired] int value) =>
        AddExpression(graphId, Expression.NewConstant(value));

    [Route("{graphId}/where")]
    public ActionResult<int> Where(string graphId, [BindRequired] int id, [BindRequired] string op, [BindRequired] int arg) =>
        AddWhereExpression(graphId, id, OperatorExtensions.Parse(op), arg);

    /// ----------------
    /// Map expressions
    /// ----------------
    [Route("{graphId}/map/not")]
    public ActionResult<int> Not(string graphId) =>
        AddMapExpression(graphId, Operator.Not, new int[] { });

    [Route("{graphId}/map/lte")]
    public ActionResult<int> LetLessThanEquals(string graphId, [BindRequired] int left, [BindRequired] int right) =>
        AddMapExpression(graphId, Operator.LessThanEquals, new int[] { left, right });

    [Route("{graphId}/map/eq")]
    public ActionResult<int> Eq(string graphId, [BindRequired] int left, [BindRequired] int right) =>
        AddMapExpression(graphId, Operator.Eq, new int[] { left, right });

    [Route("{graphId}/map/add")]
    public ActionResult<int> Add(string graphId, [BindRequired] int left, [BindRequired] int right, [BindRequired] int width) =>
        AddMapExpression(graphId, Operator.NewAdd(width), new int[] { left, right });

    [Route("{graphId}/map/multiply")]
    public ActionResult<int> Multiply(string graphId, [BindRequired] int left, [BindRequired] int right, [BindRequired] int width) =>
        AddMapExpression(graphId, Operator.NewMultiply(width), new int[] { left, right });

    [Route("{graphId}/map/if")]
    public ActionResult<int> If(string graphId, [BindRequired] int cond, [BindRequired] int t, [BindRequired] int e) =>
        AddMapExpression(graphId, Operator.If, new int[] { cond, t, e });

    [Route("{graphId}/map/id")]
    public ActionResult<int> Id(string graphId, [BindRequired] int id) =>
        AddMapExpression(graphId, Operator.Id, new int[] { id });

    [Route("{graphId}/map/gt")]
    public ActionResult<int> GreaterThan(string graphId, [BindRequired] int left, [BindRequired] int right) =>
        AddMapExpression(graphId, Operator.GreaterThan, new int[] { left, right });

    [Route("{graphId}/map/in")]
    public ActionResult<int> In(string graphId, int id, [BindRequired] string[] values) =>
        AddMapExpression(graphId, Operator.NewIn(ListModule.OfSeq(values.Select(id => int.Parse(id.Trim())))), new int[] { id });

    [Route("{graphId}/map/and")]
    public ActionResult<int> And(string graphId, [BindRequired] int left, [BindRequired] int right) =>
        AddMapExpression(graphId, Operator.And, new int[] { left, right });

    [Route("{graphId}/map/or")]
    public ActionResult<int> Or(string graphId, [BindRequired] int left, [BindRequired] int right) =>
        AddMapExpression(graphId, Operator.Or, new int[] { left, right });

    private ActionResult<int> AddExpression(string graphId, Expression expression)
    {
        if (_graphs.TryFind(graphId, out var graph))
        {
            return graph.Add(expression);
        }

        return NotFound();
    }

    private ActionResult<int> AddMapExpression(string graphId, [BindRequired] Operator op, int[] argIds)
    {
        if (_graphs.TryFind(graphId, out var graph))
        {
            var kets = ListModule.OfSeq(argIds.Select((k, idx) => graph[k]));
            var expression = Expression.NewMap(op, kets);
            return graph.Add(expression);
        }

        return NotFound();
    }

    private ActionResult<int> AddWhereExpression(string graphId, int id, Operator op, int arg)
    {
        if (_graphs.TryFind(graphId, out var graph))
        {
            var target = graph[id];
            var args = ListModule.OfSeq(new KetValue[] { graph[arg] });
            var expression = Expression.NewWhere(target, op, args);
            return graph.Add(expression);
        }

        return NotFound();
    }

    [Route("{graphId}/~sample")]
    public IActionResult Sample(string graphId, [BindRequired] string ids, int filterId, string backend = "classic")
    {
        if (_graphs.TryFind(graphId, out var graph))
        {
            var ctx = new aleph.kets.PrepareContext(qpu: GetBackend(backend));
            var kets = GetKets(graph, ids);

            var result = (filterId > 0)
                ? sample_when(ctx, kets, graph[filterId])
                : sample(ctx, kets);

            if (result.IsError)
            {
                return new StatusCodeResult(500);
            }
            else
            {
                return Ok(result.ResultValue.ToArray());
            }
        }

        return NotFound();
    }


    [HttpGet("{graphId}/~prepare")]
    public IActionResult Prepare(string graphId, [BindRequired] string ids, int filter = -1, string backend = "classic")
    {
        if (_graphs.TryFind(graphId, out var graph))
        {
            var ctx = new aleph.kets.PrepareContext(qpu: GetBackend(backend));
            var kets = GetKets(graph, ids);

            var result = (filter > 0)
                ? prepare_when(ctx, kets, graph[filter])
                : prepare(ctx, kets);

            if (result.IsError)
            {
                return new StatusCodeResult(500);
            }
            else
            {
                return Ok(result.ResultValue);
            }
        }

        return NotFound();
    }

    [Route("{graphId}/~csv")]
    [Produces("text/plain")]
    public async Task<IActionResult> Csv(string graphId, [BindRequired] string ids, int filter = -1, string backend = "classic")
    {
        if (_graphs.TryFind(graphId, out var graph))
        {
            var ctx = new aleph.kets.PrepareContext(qpu: GetBackend(backend));
            var kets = GetKets(graph, ids);

            var result = (filter > 0)
                ? prepare_when(ctx, kets, graph[filter])
                : prepare(ctx, kets);

            if (result.IsError)
            {
                return new StatusCodeResult(500);
            }
            else
            {
                return await ToCSV(graph, result.ResultValue);
            }
        }

        return NotFound();
    }

    private FSharpList<KetValue> GetKets(QuantumGraph graph, string ids) =>
        ListModule.OfSeq(
                    ids
                    .Split(',')
                    .Select(id => int.Parse(id.Trim()))
                    .Select(k => graph[k]));

    private QPU GetBackend(string backend)
    {
        if (backend == "classic")
        {
            return new aleph.qpu.classic.Processor();
        }
        else if (backend == "sparse" || backend == "qsharp")
        {
            return new aleph.qpu.qsharp.Processor(new SparseSimulator());
        }
        else
        {
            throw new ArgumentException("Unknown backend: " + backend);
        }
    }

    private bool IsMapId(KetValue ket)
    {
        if (ket.Expression.IsMap)
        {
            var map = (Expression.Map)ket.Expression;
            return map.op.IsId;
        }
        else if (ket.Expression.IsWhere)
        {
            var map = (Expression.Where)ket.Expression;
            return map.clause.IsId;
        }

        return false;
    }

    private async Task<IActionResult> ToCSV(QuantumGraph graph, IUniverse u)
    {
        var universe = (aleph.qpu.classic.Universe)u;
        System.IO.Pipelines.PipeWriter writer = Response.BodyWriter;
        var encoding = new System.Text.UTF8Encoding(false);

        var headers = universe.Allocations
            .OrderBy(kv => kv.Value)
            .Select(kv => kv.Key)
            .Where(k => !IsMapId(graph[k]))
            .ToArray();

        await writer.WriteAsync(encoding.GetBytes(string.Join(',', headers) + "\n"));

        foreach (var row in universe.State.Rows)
        {
            await writer.WriteAsync(encoding.GetBytes(string.Join(",", row) + "\n"));
        }

        await writer.CompleteAsync();

        return new EmptyResult();
    }
}

File: src/server/GraphController.fs
namespace aleph.server

open System
open Microsoft.AspNetCore.Mvc
open Microsoft.Extensions.Logging

open aleph.kets
open aleph.utils
open Utils

[<ApiController>]
[<Route("[controller]")>]
type GraphController (logger : ILogger<GraphController>, graphs: IGraphsService) =
    inherit ControllerBase()

    let getKets (graph: QuantumGraph) (ids: string) =
        ids.Split(',')
        |> Array.map (fun id -> id |> int)
        |> Array.map (fun k -> graph[k])
        |> Array.toList

    [<HttpGet>]
    member this.Get() =
        ":ℵ-0.9:" |> this.Ok

    [<HttpGet("~create")>]
    [<HttpPost("")>]
    member _.Create() =
        let id = Guid.NewGuid().ToString("D")
        let g = QuantumGraph()
        graphs.Add(id, g) |> ignore
        id

    [<Route("{graphId}")>]
    member this.GetGraph(graphId: string) =
        match graphs.TryFind(graphId) with
        | Some g -> g |> this.Ok :> IActionResult
        | _ -> this.NotFound()

    [<Route("{graphId}/{ketid}")>]
    member this.GetNode(graphId: string, ketId: int) : IActionResult =
        match graphs.TryFind(graphId) with
        | Some graph -> this.Ok(GraphNode(ketId, graph))
        | _ -> this.NotFound()

    [<Route("{graphId}/literal")>]
    member this.Literal(graphId: string, width: int) =
        this.AddExpression(graphId, Expression.Literal width)

    [<Route("{graphId}/bool")>]
    member this.ConstantBool(graphId: string, value: bool) =
        this.AddExpression(graphId, Expression.Constant (if value then 1 else 0))

    [<Route("{graphId}/int")>]
    member this.ConstantInt(graphId: string, value: int) =
        this.AddExpression(graphId, Expression.Constant value)

    [<Route("{graphId}/where")>]
    member this.Where(graphId: string, id: int, op: string, arg: int) =
        this.AddWhereExpression(graphId, id, Operator.Parse(op), arg)

    [<Route("{graphId}/where-in")>]
    member this.WhereIn(graphId: string, id: int, values: string) =
        this.AddWhereExpression(graphId, id, Operator.In (values.Split(',') |> Array.map int |> Array.toList), id)

    [<Route("{graphId}/map/not")>]
    member this.Not(graphId: string, id: int) =
        this.AddMapExpression(graphId, Operator.Not, [ id ])

    [<Route("{graphId}/map/lte")>]
    member this.LessThanEquals(graphId: string, left: int, right: int) =
        this.AddMapExpression(graphId, Operator.LessThanEquals, [ left; right ])

    [<Route("{graphId}/map/eq")>]
    member this.Eq(graphId: string, left: int, right: int) =
        this.AddMapExpression(graphId, Operator.Eq, [ left; right ])

    [<Route("{graphId}/map/add")>]
    member this.Add(graphId: string, left: int, right: int, width: int) =
        this.AddMapExpression(graphId, Operator.Add width, [ left; right ])

    [<Route("{graphId}/map/multiply")>]
    member this.Multiply(graphId: string, left: int, right: int, width: int) =
        this.AddMapExpression(graphId, Operator.Multiply width, [ left; right ])

    [<Route("{graphId}/map/id")>]
    member this.Id(graphId: string, id: int) =
        this.AddMapExpression(graphId, Operator.Id, [ id ])

    [<Route("{graphId}/map/in")>]
    member this.In(graphId: string, id: int, [<FromQuery>] value: int[]) =
        this.AddMapExpression(graphId, Operator.In (value |> Array.toList), [ id ])

    [<Route("{graphId}/map/gt")>]
    member this.GreaterThan(graphId: string, left: int, right: int) =
        this.AddMapExpression(graphId, Operator.GreaterThan, [ left; right ])

    [<Route("{graphId}/map/and")>]
    member this.And(graphId: string, left: int, right: int) =
        this.AddMapExpression(graphId, Operator.And, [ left; right ])

    [<Route("{graphId}/map/or")>]
    member this.Or(graphId: string, left: int, right: int) =
        this.AddMapExpression(graphId, Operator.Or, [ left; right ])

    member this.AddExpression(graphId: string, expression: Expression) =
        match graphs.TryFind(graphId) with
        | Some graph -> graph.Add(expression) |> this.Ok :> IActionResult
        | _ -> this.NotFound()

    member private this.AddWhereExpression(graphId: string, id: int, op: Operator, arg: int) =
        match graphs.TryFind(graphId) with
        | Some graph ->
            let target = graph.[id]
            let args = [ graph.[arg] ]
            let expression = Expression.Where(target, op, args)
            graph.Add(expression) |> this.Ok :> IActionResult
        | _ -> this.NotFound()

    member private this.AddMapExpression(graphId: string, op: Operator, argIds: int list) =
        match graphs.TryFind(graphId) with
        | Some graph ->
            let kets = argIds |> List.map (fun k -> graph.[k])
            let expression = Expression.Map(op, kets)
            graph.Add(expression) |> this.Ok :> IActionResult
        | _ -> this.NotFound()

    [<Route("{graphId}/~sample")>]
    member this.Sample(graphId: string, ids: string, filter: int) = //, backend: Backend option
        match graphs.TryFind(graphId) with
        | Some graph ->
            let ctx = this.getQuantumContext()
            let kets = getKets graph ids

            let result =
                if filter > 0 then
                    sample_when ctx (kets, graph.[filter])
                else
                    sample ctx kets

            match result with
            | Ok value -> value |> this.Ok :> IActionResult
            | _ -> new StatusCodeResult(500) :> IActionResult
        | _ -> this.NotFound() :> IActionResult

    [<Route("{graphId}/~prepare")>]
    member this.Prepare(graphId: string, ids: string, filter: int) =
        match graphs.TryFind(graphId) with
        | Some graph ->
            let ctx = this.getQuantumContext()
            let kets = getKets graph ids

            let result =
                if filter > 0 then
                    prepare_when ctx (kets, graph.[filter])
                else
                    prepare ctx kets

            match result with
            | Ok value -> value |> this.Ok :> IActionResult
            | _ -> new StatusCodeResult(500) :> IActionResult
        | _ -> this.NotFound() :> IActionResult


    [<Route("{graphId}/~histogram")>]
    member this.Histogram(graphId: string, ids: string, filter: int) : IActionResult=
        match graphs.TryFind(graphId) with
        | Some graph ->
            let prep = this.Prepare(graphId, ids, filter)
            match prep with
            | :? OkObjectResult as okResult ->
                let universe = okResult.Value :?> IUniverse
                let kets = getKets graph ids
                
                match universe.Histogram(kets) with
                | Result.Ok h -> 
                    let map_key (key: int list) = key |> List.map string |> String.concat ","
                    let map_values (m: Map<string, int>) key value = m.Add(map_key key, value)
                    h |> Map.fold map_values Map.empty |> this.Ok :> IActionResult
                | Result.Error _ -> new StatusCodeResult(500)
            | _ -> prep
        | _ -> this.NotFound() :> IActionResult


    member private this.getQuantumContext() = 
        if this.Request.Query.Keys.Contains("backend") then
            let b = this.Request.Query["backend"].Item 0
            match b.ToLowerInvariant() with
            | "classic" -> 
                { qpu = aleph.qpu.classic.Processor() }
            | "quantum" | "sparsesimulator" ->
                { qpu = aleph.qpu.qsharp.Processor(aleph.qpu.qsharp.context.simulator) }
            | _ -> failwith $"Invalid backend: {b}"
        else 
            { qpu = aleph.qpu.classic.Processor() }